android/src/main/java/com/tailscale/ipn/ui/view/SettingsView.kt

```diff
@@ -45,81 +44,104 @@ fun SettingsView(
           }
         }
       }
 }
 
 @Composable
 fun SettingRow(setting: Setting) {
-  val enabled = setting.enabled.collectAsState().value
-  val swVal = setting.isOn?.collectAsState()?.value ?: false
-  val txtVal = setting.value?.collectAsState()?.value
-
   Box {
     when (setting.type) {
-      SettingType.TEXT ->
-          ListItem(
-              modifier = Modifier.clickable { if (enabled) setting.onClick() },
-              headlineContent = {
-                Text(
-                    setting.title.getString(),
-                    style = MaterialTheme.typography.bodyMedium,
-                    color =
-                        if (setting.destructive) ts_color_dark_desctrutive_text
-                        else MaterialTheme.colorScheme.primary)
-              },
-          )
-      SettingType.SWITCH ->
-          ListItem(
-              modifier = Modifier.clickable { if (enabled) setting.onClick() },
-              headlineContent = { Text(setting.title.getString()) },
-              trailingContent = {
-                TintedSwitch(checked = swVal, onCheckedChange = setting.onToggle, enabled = enabled)
-              })
+      SettingType.TEXT -> TextRow(setting)
+      SettingType.SWITCH -> SwitchRow(setting)
       SettingType.NAV -> {
-        ListItem(
-            modifier = Modifier.clickable { if (enabled) setting.onClick() },
-            headlineContent = {
-              Text(
-                  setting.title.getString(),
-                  style = MaterialTheme.typography.bodyMedium,
-                  color =
-                      if (setting.destructive) ts_color_dark_desctrutive_text
-                      else MaterialTheme.colorScheme.primary)
-            },
-            supportingContent = {
-              txtVal?.let { Text(text = it, style = MaterialTheme.typography.bodyMedium) }
-            })
+        NavRow(setting)
       }
     }
   }
 }
 
+@Composable
+private fun TextRow(setting: Setting) {
+  val enabled = setting.enabled.collectAsState().value
+  ListItem(
+      modifier = Modifier.clickable { if (enabled) setting.onClick() },
+      headlineContent = {
+        Text(
+            setting.title ?: stringResource(setting.titleRes),
+            style = MaterialTheme.typography.bodyMedium,
+            color =
+                if (setting.destructive) ts_color_dark_desctrutive_text
+                else MaterialTheme.colorScheme.primary)
+      },
+  )
+}
+
+@Composable
+private fun SwitchRow(setting: Setting) {
+  val enabled = setting.enabled.collectAsState().value
+  val swVal = setting.isOn?.collectAsState()?.value ?: false
+  ListItem(
+      modifier = Modifier.clickable { if (enabled) setting.onClick() },
+      headlineContent = { Text(setting.title ?: stringResource(setting.titleRes)) },
+      trailingContent = {
+        TintedSwitch(checked = swVal, onCheckedChange = setting.onToggle, enabled = enabled)
+      })
+}
+
+@Composable
+private fun NavRow(setting: Setting) {
```

Comment:
Each setting type has its own function now so that we only collectAsState what we need for that type. That's particularly useful for NavRow since it has no dynamic state.